File: app/src/main/java/com/akari/ppx/xp/hook/code/assist/CopyHook.java
package com.akari.ppx.xp.hook.code.assist;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.ClipData;
import android.content.ClipboardManager;

import com.akari.ppx.common.utils.Utils;
import com.akari.ppx.common.utils.XSP;
import com.akari.ppx.xp.hook.code.SuperbHook;

import java.util.HashMap;

import de.robv.android.xposed.XC_MethodHook;
import de.robv.android.xposed.XC_MethodReplacement;

import static com.akari.ppx.common.constant.Prefs.COPY_ITEM;
import static de.robv.android.xposed.XposedHelpers.callMethod;
import static de.robv.android.xposed.XposedHelpers.findClass;

public class CopyHook extends SuperbHook {
	private final String OptionActionType = "ACTION_PI";

	@SuppressLint("WrongConstant")
	private void copyText(ClassLoader cl, Object activity, Object absFeedCell) {
		String text;
		try {
			text = (String) callMethod(callMethod(absFeedCell, "getFeedItem"), "getContent");
		} catch (Error ignored) {
			text = (String) callMethod(callMethod(absFeedCell, "getComment"), "getText");
		}
		((ClipboardManager) ((Activity) activity).getSystemService("clipboard")).setPrimaryClip(ClipData.newPlainText(text, text));
		Utils.showSystemToastXP(cl, "复制成功");
	}

	@Override
	protected void onHook(ClassLoader cl) {
		if (!XSP.get(COPY_ITEM)) return;
		hookMethod("com.sup.android.mi.feed.repo.utils.AbsFeedCellUtil$Companion", "canBeRecreated", "com.sup.android.mi.feed.repo.bean.cell.AbsFeedCell", XC_MethodReplacement.returnConstant(true));
		hookMethod("android.view.View", "setTag", Object.class, new XC_MethodHook() {
			@Override
			protected void beforeHookedMethod(MethodHookParam param) {
				try {
					if (Enum.valueOf((Class) findClass("com.sup.android.i_sharecontroller.model.OptionAction$OptionActionType", cl), OptionActionType).equals(param.args[0])) {
						callMethod(callMethod(param.thisObject, "getChildAt", 1), "setText", "复制文字");
					}
				} catch (Error ignored) {
				}
			}
		});
		hookMethod("com.sup.android.base.feed.a", "a", Activity.class, "com.sup.android.mi.feed.repo.bean.cell.AbsFeedCell", String.class, String.class, String.class, boolean.class, new XC_MethodReplacement() {
			@Override
			protected Object replaceHookedMethod(MethodHookParam param) {
				copyText(cl, param.args[0], param.args[1]);
				return null;
			}
		});
		hookMethod("com.sup.android.base.e.a", "a", Activity.class, "com.sup.android.mi.feed.repo.bean.cell.AbsFeedCell", String.class, String.class, String.class, HashMap.class, boolean.class, new XC_MethodReplacement() {
			@Override
			protected Object replaceHookedMethod(MethodHookParam param) {
				copyText(cl, param.args[0], param.args[1]);
				return null;
			}
		});
		hookMethod("kotlin.jvm.internal.Intrinsics", "throwUninitializedPropertyAccessException", String.class, XC_MethodReplacement.DO_NOTHING);
	}
}

File: app/src/main/java/com/akari/ppx/xp/hook/assist/CopyHook.kt
@file:Suppress("unused", "unchecked_cast", "type_mismatch_warning")

package com.akari.ppx.xp.hook.assist

import android.app.Activity
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.view.View
import de.robv.android.xposed.XC_MethodHook
import com.akari.ppx.utils.*
import com.akari.ppx.xp.Init.absFeedCellClass
import com.akari.ppx.xp.Init.cl
import com.akari.ppx.xp.Init.enterPi1
import com.akari.ppx.xp.Init.enterPi1Class
import com.akari.ppx.xp.Init.enterPi2
import com.akari.ppx.xp.Init.enterPi2Class
import com.akari.ppx.xp.hook.SwitchHook
import java.lang.Enum.valueOf

class CopyHook : SwitchHook("copy_item") {
    override fun onHook() {
        val actionType = valueOf(
            "com.sup.android.i_sharecontroller.model.OptionAction\$OptionActionType".findClass(cl) as Class<Enum<*>>,
            "ACTION_PI"
        )
        "com.sup.android.mi.feed.repo.utils.AbsFeedCellUtil\$Companion".replaceMethod(
            cl,
            "canBeRecreated",
            absFeedCellClass
        ) { true }
        View::class.java.name.hookBeforeMethod(cl, "setTag", Object::class.java) { param ->
            runCatching {
                param.args[0].check(actionType) {
                    param.thisObject.callMethod("getChildAt", 1)?.callMethod("setText", "复制文字")
                }
            }
        }

        fun XC_MethodHook.MethodHookParam.copyText() {
            val text: String? = runCatching {
                args[1].callMethod("getFeedItem")?.callMethodAs<String>("getContent")
            }.getOrElse {
                args[1].callMethod("getComment")?.callMethodAs<String>("getText")
            }
            ((args[0] as Activity).getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager).setPrimaryClip(
                ClipData.newPlainText(text, text)
            )
            showStickyToast("复制成功")
        }

        enterPi1Class!!.replaceMethod(
            enterPi1(),
            Activity::class.java,
            absFeedCellClass,
            String::class.java,
            String::class.java,
            String::class.java,
            Boolean::class.java
        ) { param ->
            param.copyText()
        }
        enterPi2Class!!.replaceMethod(
            enterPi2(),
            Activity::class.java,
            absFeedCellClass,
            String::class.java,
            String::class.java,
            String::class.java,
            HashMap::class.java,
            Boolean::class.java
        ) { param ->
            param.copyText()
        }
        "kotlin.jvm.internal.Intrinsics".replaceMethod(
            cl,
            "throwUninitializedPropertyAccessException",
            String::class.java
        ) {}
    }
}